x/distribution/keeper/calculation.go

```diff
@@ -0,0 +1,161 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/cosmos/cosmos-sdk/x/distribution/types"
+)
+
+// initialize rewards for a new validator
+func (k Keeper) initializeValidator(ctx sdk.Context, val sdk.Validator) {
+	// set initial historical rewards (period 0)
+	k.SetValidatorHistoricalRewards(ctx, val.GetOperator(), 0, types.ValidatorHistoricalRewards{})
+
+	// set current rewards (starting at period 1)
+	k.SetValidatorCurrentRewards(ctx, val.GetOperator(), types.ValidatorCurrentRewards{
+		Rewards: sdk.DecCoins{},
+		Period:  1,
+	})
+
+	// set accumulated commission
+	k.SetValidatorAccumulatedCommission(ctx, val.GetOperator(), types.ValidatorAccumulatedCommission{})
+}
+
+// increment validator period, returning the period just ended
+func (k Keeper) incrementValidatorPeriod(ctx sdk.Context, val sdk.Validator) uint64 {
+	// Fetch current rewards
+	rewards := k.GetValidatorCurrentRewards(ctx, val.GetOperator())
+
+	// Calculate current ratio
+	var current sdk.DecCoins
+	if val.GetTokens().IsZero() {
+
+		// can't calculate ratio for zero-token validators
+		// ergo we instead add to the community pool
```

Comment:
when would this come up?

Reply:
If a delegator withdraws after a validator has been slashed to zero, for example (and there's a bit of delay in the Tendermint validator updates, so they would still sign a block or two and get rewards).

Reply:
thanks